### ISSUE TITLE
Back pressure instead of dropping spans while using `sending_queue`

### ISSUE THREAD

**Issue by tanner-bruce** [open-telemetry/opentelemetry-collector#5456]
**Is your feature request related to a problem? Please describe.**
When using a pull based receiver alongside an exporter using the parallel `sending_queue`, it is possible for the receiver to quickly overload the pipeline, and start dropping spans, if the exporter is slower than the receiver. 

**Describe the solution you'd like**
Having a setting to allow the in memory queue to block, instead of dropping spans, would allow us to apply back pressure back to the receiver itself.

**Describe alternatives you've considered**
Not using the built in sending queue and instead implementing our own parallel processing helper.

**Comment by bogdandrutu** [comment]
> Having a setting to allow the in memory queue to block, instead of dropping spans, would allow us to apply back pressure back to the receiver itself.

Why not implementing that in the receiver, check that you get an error and take the according action? We try to keep the design very simple, and ensure that the pipeline returns immediately and the "caller" will take the right action. This is because in case of a "push" receiver you don't want to block the call, but let the caller know that you are having troubles.

**Comment by tanner-bruce** [comment]
Do you mean don't use the sending queue at all and instead reimplement it in my receiver? Or check if the error returned is `errSendingQueueIsFull = errors.New("sending_queue is full")` and busy loop until it's successful?

**Comment by tanner-bruce** [comment]
Yes, we need to slow down the producer, which is inside a receiver. We are thinking to completely remove the sending queue and instead rely on parallelism from the receiver end to avoid a busy loop or a backoff retry loop, although this means we cannot use batching in the pipeline. This is fine in this case because these trace bundles are pre batched. If this doesn't work, we can roll our own queue.

We want/need to process as fast as possible (hundreds of millions of span per minute collection rate) so performing unnecessary work or extra waiting isn't great. Arguably, blocking in the queue would be simpler than the existing implementation, although it would only be applicable to pipelines using a pull-based receiver, for the reasons you've stated above.

For context, we are implementing a fan out model of collectors, with a single ingestion set which writes to pubsub, and then each of our "real" exporters will have their own deployments which will pull from the subscription, to reduce the blast radius of any given export pipeline from the others.
